File: Nololiyt.IndexedDbSharp.Raw/Nololiyt.IndexedDbSharp.Raw/Typescripts/Async/Entities/IdbKeyRangeInfo.ts
export type IdbKeyRangeBoundMode = "bound" | "lowerBound" | "upperBound";

export interface IdbKeyRangeInfo
{
    lower: any;
    lowerOpen: boolean;
    upper: any;
    upperOpen: boolean;
    boundMode: IdbKeyRangeBoundMode;
}

export function isIdbKeyRangeInfo(obj: any): obj is IdbKeyRangeInfo
{
    const converted = <IdbKeyRangeInfo>obj;
    return converted.boundMode !== undefined &&
        converted.lowerOpen !== undefined &&
        converted.upperOpen !== undefined;
}

export function convertToIdbKeyRange(info: IdbKeyRangeInfo): IDBKeyRange
{
    switch (info.boundMode)
    {
        case "lowerBound":
            return IDBKeyRange.lowerBound(info.lower, info.lowerOpen);
        case "upperBound":
            return IDBKeyRange.upperBound(info.upper, info.upperOpen);
        default:
            return IDBKeyRange.bound(info.lower, info.upper, info.lowerOpen, info.upperOpen);
    }
}

File: Nololiyt.IndexedDbSharp.Raw/Nololiyt.IndexedDbSharp.Raw/CSharp/Entities/IdbKeyRangeInfo.cs
using Microsoft.Extensions.Logging;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Nololiyt.IndexedDbSharp.Raw.CSharp.Entities
{
    public sealed record IdbKeyRangeInfo(
        IdbKeyRangeBoundMode BoundMode, 
        object? Lower, bool LowerOpen,
        object? Upper, bool UpperOpen)
    {
        public static IdbKeyRangeInfo LowerBound<T>(T lower, bool open = false)
        {
            return new IdbKeyRangeInfo(
                IdbKeyRangeBoundMode.LowerBound,
                lower, open,
                default, false);
        }
        public static IdbKeyRangeInfo UpperBound<T>(T upper, bool open = false)
        {
            return new IdbKeyRangeInfo(
                IdbKeyRangeBoundMode.UpperBound,
                default, false,
                upper, open);
        }
        public static IdbKeyRangeInfo Bound<TLower, TUpper>(
            TLower lower, TUpper upper, bool lowerOpen = false, bool upperOpen = false)
        {
            return new IdbKeyRangeInfo(
                IdbKeyRangeBoundMode.Bound,
                lower, lowerOpen,
                upper, upperOpen);
        }
        public static IdbKeyRangeInfo Only<T>(T value)
        {
            return new IdbKeyRangeInfo(
                IdbKeyRangeBoundMode.Bound,
                value, false,
                value, false);
        }
    }
}
